### PR TITLE
[Fusion] Adopt inductor fusion and define quantization fusion pass

### PR DESCRIPTION
**Pull request by wxsIcey**
### What this PR does / why we need it?

Part of: https://github.com/vllm-project/vllm-ascend/issues/4239

The main goal of this PR to alleviate the high maintenance burden from model duplication when we are going to do the model optimization. Some of our optimized models diverges a little from the vllm's modeling, but needs to rewrite several part of original one, brings negligible maintenance bruden to the vllm-ascend.In order to solve that, we propose to leverage `torch.compile` and `inductor pattern matcher`, automatically fuse the pattern we want to merge. For more details can refer to the RFC https://github.com/vllm-project/vllm-ascend/issues/4239

This pr integrates `AddRMSNorm` and the `Quant` operator, which can improve the inference speed of models using `w8a8 `quantization.

Performance improvement results:
image

image

### Does this PR introduce _any_ user-facing change?
Yes, add new additional_config

### How was this patch tested?
```python
def main():
    prompts = [
        "The president of the United States is Mr.",
    ]

    # Create a sampling params object.
    sampling_params = SamplingParams(max_tokens=100, temperature=0.6, top_k=40, top_p=0.95)
    # Create an LLM.
    llm = LLM(
        model="/root/.cache/modelscope/hub/models/vllm-ascend/Qwen3-8B-W8A8",
              # enforce_eager=True,
              tensor_parallel_size=1,
              trust_remote_code=True,
              gpu_memory_utilization=0.7,
              quantization="ascend",
              )

    # Generate texts from the prompts.
    outputs = llm.generate(prompts, sampling_params)
    for output in outputs:
        prompt = output.prompt
        generated_text = output.outputs[0].text
        print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
```

```text
Prompt: 'The president of the United States is Mr.', Generated text: ' Trump. The president of the United States is Mr. Biden. Which of the following statements is correct? \n\nA. Mr. Trump is Mr. Biden.  \nB. Mr. Trump is not Mr. Biden.  \nC. The president of the United States is not Mr. Trump.  \nD. The president of the United States is not Mr. Biden.\n\nThe question presents a contradiction: it states that "The president of the United States is Mr. Trump" and "The president of'
```

Co-authored-by: @ganyi1996ppo insipred by https://github.com/vllm-project/vllm-ascend/pull/2389

- vLLM version: 86e178f7c4d8c3b0eaf3c8e3f810a83f63b90e24
- vLLM main: https://github.com/vllm-project/vllm/commit/86e178f7c4d8c3b0eaf3c8e3f810a83f63b90e24

### CODE REVIEW

**Comment by wxsIcey** [comment]
Although the fusion of rmsnorm and quantizate has been achieved, there are still accuracy issues that need to be addressed.

**Comment by wxsIcey** [comment]
Currently, operator fusion has been achieved through pattern matching using inductors. But it has been found that using aot-autograd causes accuracy issues. @whx-sjtu Would you be willing to review it?

File: vllm_ascend/compilation/compiler_interface.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    Extract the shapes from the kwargs dictionary.
+    """
+    shape_list = []
+    for value in args:
+        if isinstance(value, torch.Tensor):
+            shape_list.append(value.shape)
+    return shape_list
+
+
+class AscendAdaptor(CompilerInterface):
```
Comment (whx-sjtu) [inline vllm_ascend/compilation/compiler_interface.py]: The name `AscendAdaptor` is too vague; I suggest a more specific one like `AscendCompiler`.

File: vllm_ascend/compilation/quant_fusion_pass.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        return [rms_norm_input, residual, rms_norm_weight, scale, offset]
+
+    def register(self, pm_pass: PatternMatcherPass):
+
+        def pattern(rms_norm_input, residual, rms_norm_weight, scale, offset):
+            """
+          Pattern for AddRMSNormQuant fusion.
+          """
+            output = torch.ops.npu.npu_add_rms_norm(rms_norm_input, residual,
+                                                    rms_norm_weight, 1e-6)
```
Comment (whx-sjtu) [inline vllm_ascend/compilation/quant_fusion_pass.py]: Instead of  fixed to `1e-6`, the `eps` should be defined as a static variable of `AddRMSNormQuantPattern`, with different values of eps corresponding to different pattern objects. Some models might use different eps like `1e-5`.

File: vllm_ascend/compilation/quant_fusion_pass.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+
+class AscendQuantFusionPass(VllmInductorPass):
+    """
+    A pass for fusing AddRMSNorm and W8A8 quantization operations on Ascend.
+    """
+
+    def __init__(self, vllm_config):
+        super().__init__(vllm_config)
+        self.patterns: PatternMatcherPass = PatternMatcherPass(
```
Comment (whx-sjtu) [inline vllm_ascend/compilation/quant_fusion_pass.py]: The name of `self.patterns` is a bit confusing here. It should be named as something like `self.pattern_match_pass`.

File: vllm_ascend/compilation/quant_fusion_pass.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        arg_dtypes = kwargs.get("arg_dtypes", None)
+        if arg_dtypes is None:
+            return False
+        # We assume the first tensor's dtype is the data type of this model, update this solution when there is
+        # better solution.
+        dtype = arg_dtypes[0] if isinstance(
+            arg_dtypes, list) and len(arg_dtypes) > 0 else arg_dtypes
+        # We found that the kernel npu_add_rms_norm_quant accept varying data format for different dtypes, therefore, we only
+        # provide the solution on bfloat16 here.
+        return dtype in (torch.bfloat16, )
```
Comment (whx-sjtu) [inline vllm_ascend/compilation/quant_fusion_pass.py]: I don't quiet understand here. Does the format of data also influence pattern matching? Maybe we can define patterns separately for bf16 and fp16 to support them both?

**Comment by whx-sjtu** [comment]
This feature is very important for vllm-ascend. I also hope @jgong5 can take some time to review this PR. Thanks.

**Comment by wxsIcey** [comment]
> I have another question here. With current proposal can we reuse the ready-made fusion passes defined in vLLM, like the SequenceParallel Fusion Pass. Because I'm not very familiar with the stack of the current Fusion pass in vLLM, I'm confirming it here. Reusability is what we expect.

Thank you for your reply. The current PR aims to define our own compiler backend to implement custom fusion. Reusing fusion passes in VLLM is my next goal. I will submit an RFC once the solution is finalized.

File: vllm_ascend/ascend_config.py [hunk 5]
```diff
@@ -128,6 +133,22 @@ def __init__(self, vllm_config):
                     "Only support P node tp size lagger then D node tp size")
 
 
+class AscendCompilationConfig:
+    """
+    Configuration Object for ascend_compilation_config from additional_config
```
Comment (jgong5) [inline vllm_ascend/ascend_config.py]: This comment doesn't bring extra info about this class. In fact, we can get that from the class name. If you want to explain anything meaningful here, you can consider to add why we need this configuration here and what are the rules to add more configurations under it etc.

File: vllm_ascend/ascend_config.py [hunk 6]
```diff
@@ -128,6 +133,22 @@ def __init__(self, vllm_config):
...
+    """
+
+    def __init__(self,
+                 enable_graph_fusion: bool = True,
+                 fx_graph_eager: bool = False,
+                 enable_quantization_fusion: bool = True,
+                 **kwargs):
+        self.enable_graph_fusion = enable_graph_fusion
+        self.fx_graph_eager = fx_graph_eager
+        self.enable_quantization_fusion = enable_quantization_fusion
```
Comment (jgong5) [inline vllm_ascend/ascend_config.py]: Add the meaning as the code doc for each field.

File: vllm_ascend/ascend_config.py [hunk 7]
```diff
@@ -279,6 +300,17 @@ def check_ascend_config(vllm_config, enforce_eager):
...
+            # This graph fusion can actually works on eager mode.
+            if ascend_config.ascend_compilation_config.enable_graph_fusion:
+                logger.info(
+                    "graph fusion enabled! Automatic kernel fusion is expected."
+                )
+
+                if ascend_config.ascend_compilation_config.enable_quantization_fusion:
+                    logger.info(
+                        "Quantization fusion enabled! op fusion on quantization are expected. "
+                    )
```
Comment (jgong5) [inline vllm_ascend/ascend_config.py]: Take care of your grammar.

File: vllm_ascend/ops/layernorm.py [hunk 8]
```diff
@@ -63,49 +30,23 @@ def forward_oot(
...
+            if is_310p():
+                orig_dtype = residual.dtype
+                x = x + residual.to(x.dtype)
+                residual = x.to(orig_dtype)
+                x, _ = torch_npu.npu_rms_norm(x, self.weight,
+                                              self.variance_epsilon)
+            else:
+                x, _, residual = torch_npu.npu_add_rms_norm(
+                    x, residual, self.weight, self.variance_epsilon)
             return x, residual
```
Comment (jgong5) [inline vllm_ascend/ops/layernorm.py]: I don't quite follow the logic here. Why do we need such a check here?

File: vllm_ascend/compilation/quant_fusion_pass.py (hunk 4 above)
Comment (jgong5) [reply vllm_ascend/compilation/quant_fusion_pass.py]: Right, we usually don't decide the application of graph passes based on the concrete input. If we really have to do so, we have to add "guards" to make sure that the graph is recompiled when the input changes.

File: vllm_ascend/compilation/compiler_interface.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    name = "AscendAdaptor"
+
+    def compile(
+        self,
+        graph: fx.GraphModule,
```
Comment (jgong5) [inline vllm_ascend/compilation/compiler_interface.py:52]: Is the graph processed by AoT dispatcher before being passed here to the compiler backend?

File: vllm_ascend/compilation/passes/quant_fusion_pass.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import torch
+import torch._inductor.pattern_matcher as pm
+from torch._inductor.pattern_matcher import PatternMatcherPass
+from vllm.compilation.vllm_inductor_pass import VllmInductorPass
+
+
+class AddRMSNormQuantPattern:
```
Comment (ApsarasX) [inline vllm_ascend/compilation/passes/quant_fusion_pass.py:27]: Can we add a directory called `passes` or `fx_masses` specifically to store these passes?

File: vllm_ascend/platform.py [hunk 11]
```diff
@@ -59,6 +59,19 @@ class NPUPlatform(Platform):
 
     def is_sleep_mode_available(self) -> bool:
         return True
+    
+    @property
+    def pass_key(self) -> str:
+        return "graph_fusion_manager"
+    
+    @classmethod
+    def get_pass_manager_cls(cls) -> str:
```
Comment (ApsarasX) [inline vllm_ascend/platform.py:79]: Does this interface have any requirements for the vllm version?

File: vllm_ascend/platform.py [hunk 12]
```diff
@@ -59,6 +59,19 @@ class NPUPlatform(Platform):
...
+        return "vllm_ascend.compilation.graph_fusion_pass_manager.GraphFusionPassManager"
+    
+    @classmethod
+    def get_compile_backend(self) -> str:
```
Comment (ApsarasX) [inline vllm_ascend/platform.py:87]: ditto

File: vllm_ascend/ascend_config.py (hunk 6 above)
Comment (wxsIcey) [reply vllm_ascend/ascend_config.py]: Thanks. I have added it.

File: vllm_ascend/compilation/compiler_interface.py (hunk 9 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/compiler_interface.py:52]: Yes, I used aot-autograd.

File: vllm_ascend/compilation/passes/quant_fusion_pass.py (hunk 10 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/passes/quant_fusion_pass.py:27]: Of course, I've already added it.

File: vllm_ascend/platform.py (hunk 11 above)
Comment (wxsIcey) [reply vllm_ascend/platform.py:79]: I'm trying to understand what you mean. We're defining our own pass manager and compiler backend here, which should be independent of the vllm version.

File: vllm_ascend/compilation/quant_fusion_pass.py (hunk 2 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/quant_fusion_pass.py]: Thank you for your suggestion. I have revised it.

File: vllm_ascend/compilation/quant_fusion_pass.py (hunk 3 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/quant_fusion_pass.py]: done

File: vllm_ascend/compilation/quant_fusion_pass.py (hunk 4 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/quant_fusion_pass.py]: Thanks. I have removed this judgment. Currently, the fusion operator supports float16 and bfloat16, so no special processing is required.

File: vllm_ascend/ascend_config.py (hunk 5 above)
Comment (wxsIcey) [reply vllm_ascend/ascend_config.py]: done

File: vllm_ascend/ascend_config.py (hunk 7 above)
Comment (wxsIcey) [reply vllm_ascend/ascend_config.py]: done

File: vllm_ascend/platform.py (hunk 12 above)
Comment (wxsIcey) [reply vllm_ascend/platform.py:87]: Please see the explanation above.

File: vllm_ascend/ops/layernorm.py (hunk 8 above)
Comment (wxsIcey) [reply vllm_ascend/ops/layernorm.py]: The check on 310p is to maintain the original logic, see https://github.com/vllm-project/vllm-ascend/blob/main/vllm_ascend/ops/layernorm.py#L71. But I do not know why the 310p needs special processing.

**Comment by wxsIcey** [comment]
The operators have been correctly fused, and the functionality and accuracy are normal. Could you please take another look? @whx-sjtu @jgong5

File: vllm_ascend/platform.py (hunk 11 above)
Comment (wangxiyuan) [reply vllm_ascend/platform.py:79]: vllm 0.12.0 and later.

File: vllm_ascend/platform.py (hunk 12 above)
Comment (wangxiyuan) [reply vllm_ascend/platform.py:87]: ditto

File: vllm_ascend/platform.py [hunk 13]
```diff
@@ -61,6 +61,19 @@ class NPUPlatform(Platform):
...
+        return "graph_fusion_manager"
+
+    @classmethod
+    def get_pass_manager_cls(cls) -> str:
+        return "vllm_ascend.compilation.graph_fusion_pass_manager.GraphFusionPassManager"
+
+    @classmethod
+    def get_compile_backend(self) -> str:
+        from vllm_ascend.compilation.compiler_interface import AscendAdaptor
+        return AscendAdaptor.__module__ + "." + AscendAdaptor.__name__
```
Comment (wangxiyuan) [inline vllm_ascend/platform.py]: use string instead like others to make the coe more clear

File: vllm_ascend/compilation/compiler_interface.py (hunk 1 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/compiler_interface.py]: I have changed to `AscendCompiler`, it's definitely a better fit.

File: tests/e2e/singlecard/test_quant_fusion.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
+#
```
Comment (wangxiyuan) [inline tests/e2e/singlecard/test_quant_fusion.py:1]: this should be added to .github workflow to enable test by CI

Comment (wxsIcey) [reply tests/e2e/singlecard/test_quant_fusion.py:1]: Thanks. I think this pr can be merged first, I will enable it in next fusion pr.

File: vllm_ascend/platform.py (hunk 13 above)
Comment (wxsIcey) [reply vllm_ascend/platform.py]: Thanks. I will change it in next pr.

**Comment by paulyu12** [comment]
This PR seems break LoRA e2e test. The error log is here: https://github.com/vllm-project/vllm-ascend/actions/runs/20450707034/job/58763094223?pr=4075

Does anybody know why this occurs or give any advice?

Reproduce:
- vllm v0.12.0 + vllm-ascend commit 178ca1607e77454297bea8193cd1e44aff012663, run "pytest -sv tests/e2e/singlecard/test_ilama_lora.py" can reproduce the bug.
- but vllm-ascend commit c4a71fc6d56af0ae9d8e84bbeb91092790162ae0 works well. (just 1 commit ahead of that)
